File: node/src/main/java/dev/httpmarco/polocloud/node/commands/specific/PlatformArgument.java
package dev.httpmarco.polocloud.node.commands.specific;

import dev.httpmarco.polocloud.node.Node;
import dev.httpmarco.polocloud.node.commands.CommandArgument;
import dev.httpmarco.polocloud.node.commands.CommandContext;
import dev.httpmarco.polocloud.node.platforms.Platform;
import dev.httpmarco.polocloud.node.platforms.PlatformService;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Arrays;
import java.util.List;

public final class PlatformArgument extends CommandArgument<Platform> {

    public PlatformArgument(String key) {
        super(key);
    }

    @Override
    public boolean predication(@NotNull String rawInput) {
        return Node.instance().platformService().find(rawInput) != null;
    }

    @Override
    public List<String> defaultArgs(CommandContext context) {
        return Node.instance().platformService().platforms().stream().map(Platform::id).toList();
    }

    @Contract(pure = true)
    @Override
    public @Nullable Platform buildResult(String input) {
        return Node.instance().platformService().find(input);
    }
}

File: agent/src/main/java/dev/httpmarco/polocloud/agent/runtime/local/terminal/commands/type/PlatformArgument.kt
package dev.httpmarco.polocloud.agent.runtime.local.terminal.commands.type

import dev.httpmarco.polocloud.agent.Agent
import dev.httpmarco.polocloud.agent.runtime.local.terminal.commands.CommandArgument
import dev.httpmarco.polocloud.agent.runtime.local.terminal.commands.CommandContext
import dev.httpmarco.polocloud.platforms.Platform
import dev.httpmarco.polocloud.platforms.PlatformPool

class PlatformArgument(key: String = "platform") : CommandArgument<Platform>(key) {

    override fun buildResult(input: String, context: CommandContext): Platform {
        return PlatformPool.find(input)!!
    }

    override fun defaultArgs(context: CommandContext): MutableList<String> {
        return PlatformPool.platforms().stream().map { it.name }.toList()
    }
}